helpers/stable_swap_math/mod.rs

```diff
@@ -0,0 +1,934 @@
+    token_in_amount: u128,
+    token_out_idx: usize,
+    current_reserves: &Vec<u128>,
+    fees: &Fees,
+    amp_coef: u128,
+) -> Result<(u128, u128), MathError> {
+    let r_token_in_amount = amount_to_rated(token_in_amount, rates[token_in_idx])?;
+    let r_current_reserves = amounts_to_rated(current_reserves, rates)?;
+
+    let (r_amount_swapped, r_fee) = swap_to(
+        token_in_idx,
+        r_token_in_amount,
+        token_out_idx,
+        &r_current_reserves,
+        fees,
+        amp_coef,
+    )?;
+
+    let amount_swapped = amount_from_rated(r_amount_swapped, rates[token_out_idx])?;
+    let fee = amount_from_rated(r_fee, rates[token_out_idx])?;
+    Ok((amount_swapped, fee))
+}
+
+/// Compute SwapResult after an exchange given `amount_out` of the `token_out_id`
+/// panics if token ids are out of bounds
+/// NOTICE: it does not check if `token_in_id` != `token_out_id`
+/// /// Returns (amount_in, fee_amount)
+pub fn swap_from(
+    token_out_idx: usize,
+    token_out_amount: u128, // Net amount (w/o fee)
+    token_in_idx: usize,
+    current_reserves: &Vec<u128>,
+    fees: &Fees,
+    amp_coef: u128,
+) -> Result<(u128, u128), MathError> {
+    // fees are applied to "token_out" amount
+    let fee = fees.trade_fee_from_net(token_out_amount)?;
+    let token_out_amount_plus_fee = token_out_amount
+        .checked_add(fee)
+        .ok_or(MathError::AddOverflow(11))?;
+
+    let y = compute_y(
+        current_reserves[token_out_idx]
+            .checked_sub(token_out_amount_plus_fee)
+            .ok_or(MathError::SubUnderflow(12))?,
+        current_reserves,
+        token_out_idx,
+        token_in_idx,
+        amp_coef,
+    )?;
+    // add 1 in case there are any rounding errors
+    // https://github.com/curvefi/curve-contract/blob/b0bbf77f8f93c9c5f4e415bce9cd71f0cdee960e/contracts/pool-templates/base/SwapTemplateBase.vy#L466
+    let dy: u128 = y
+        .checked_sub(current_reserves[token_in_idx])
+        .ok_or(MathError::SubUnderflow(13))?
+        .checked_add(1)
+        .ok_or(MathError::AddOverflow(12))?;
+
+    Ok((dy, fee))
+}
+
+pub fn rated_swap_from(
+    rates: &[u128],
+    token_out_idx: usize,
+    token_out_amount: u128,
+    token_in_idx: usize,
+    current_reserves: &Vec<u128>,
+    fees: &Fees,
+    amp_coef: u128,
+) -> Result<(u128, u128), MathError> {
+    let r_token_out_amount = amount_to_rated(token_out_amount, rates[token_out_idx])?;
+    let r_current_reserves = amounts_to_rated(current_reserves, rates)?;
+    let (r_dy, r_fee) = swap_from(
+        token_out_idx,
+        r_token_out_amount,
+        token_in_idx,
+        &r_current_reserves,
+        fees,
+        amp_coef,
+    )?;
+
+    let dy = amount_from_rated(r_dy, rates[token_in_idx])?;
+    let fee = amount_from_rated(r_fee, rates[token_out_idx])?;
+    Ok((dy, fee))
+}
+
+/// Compute the amount of LP tokens to mint after a deposit
+/// return <lp_amount_to_mint, lp_fees_part>
+pub fn compute_lp_amount_for_deposit(
```

Comment:
This doesn't need to be `pub` either. It'd be good to go through all of the methods and downgrade their visibility if possible.

Reply:
Done in [RM pub where unnecessary](https://github.com/Cardinal-Cryptography/common-amm-stable-swap/pull/5/commits/27df09c7891ee7a64f85b73e3da569b976652311).